client/main.js

```diff
@@ -0,0 +1,470 @@
+import BN from "bn.js";
+import web3Utils from "web3-utils";
+
+const ganache = require("ganache-core");
+const contract = require("truffle-contract");
+const ReputationMiningCycleJSON = require("../build/contracts/ReputationMiningCycle.json");
+const ColonyNetworkJSON = require("../build/contracts/IColonyNetwork.json");
+const PatriciaTreeJSON = require("../build/contracts/PatriciaTree.json");
+
+const ReputationMiningCycle = contract(ReputationMiningCycleJSON);
+
+const jsonfile = require("jsonfile");
+
+const file = "./reputations.json";
+
+const Web3 = require("web3");
+
+const ColonyNetwork = contract(ColonyNetworkJSON);
+const accountAddress = "0xbb46703786c2049d4d6dd43f5b4edf52a20fefe4";
+
+class ReputationMiningClient {
+  constructor(minerAddress, realProviderPort = 8545) {
+    this.web3 = new Web3();
+    this.minerAddress = minerAddress;
+    this.PatriciaTree = contract(PatriciaTreeJSON);
+    this.ganacheProvider = ganache.provider({
+      network_id: 515,
+      vmErrorsOnRPCResponse: false,
+      locked: false,
+      verbose: true,
+      accounts: [
+        {
+          balance: "0x10000000000000000000000000",
+          secretKey: "0xe5c050bb6bfdd9c29397b8fe6ed59ad2f7df83d6fd213b473f84b489205d9fc7"
+        }
+      ]
+    });
+    this.PatriciaTree.setProvider(this.ganacheProvider);
+    this.realProvider = new Web3.providers.HttpProvider(`http://localhost:${realProviderPort}`);
+    ColonyNetwork.setProvider(this.realProvider);
+    ReputationMiningCycle.setProvider(this.realProvider);
+    try {
+      this.reputations = jsonfile.readFileSync(file);
+    } catch (err) {
+      this.reputations = {};
+    }
+  }
+
+  async initialise(colonyNetworkAddress) {
+    this.reputationTree = await this.PatriciaTree.new({ from: accountAddress, gas: 4000000 });
+    this.nReputations = 0;
+    this.setColonyNetworkAddress(colonyNetworkAddress);
+  }
+
+  async setColonyNetworkAddress(address) {
+    this.colonyNetwork = ColonyNetwork.at(address);
+  }
+
+  async addLogContentsToReputationTree() {
+    const makeJustificationTree = true;
+    // Snapshot the current state, in case we get in to a dispute, and have to roll back
+    // to generated the justification tree.
+    let justUpdatedProof = { value: this.getValueAsBytes(0, 0), branchMask: 0, siblings: [] };
+    let nextUpdateProof = { value: this.getValueAsBytes(0, 0), branchMask: 0, siblings: [] };
+
+    // await this.snapshotTree();
+    this.justificationTree = await this.PatriciaTree.new({ from: accountAddress, gas: 4000000 });
+    this.justificationHashes = {};
+
+    let nLogEntries = await this.colonyNetwork.getReputationUpdateLogLength(false);
+    nLogEntries = new BN(nLogEntries.toString());
+    let interimHash;
+    let jhLeafValue;
+    for (let i = new BN("0"); i.lt(nLogEntries); i.iadd(new BN("1"))) {
+      interimHash = await this.reputationTree.getRootHash(); // eslint-disable-line no-await-in-loop
```

Comment:
To avoid awaiting in loops (and also increase performance) we should consider using something like https://github.com/caolan/async. With that and splitting up this function into smaller functional parts the code will be more performant and much easier to parse.

Reply:
I don't believe we can unwind these loops by using `async`. The log entries must be applied in the same order they are in the log, which we have no guarantee of that if we fire them all off at the same time.

Reply:
`async` isn't only about doing things in parallel: https://caolan.github.io/async/docs.html#whilst

Reply:
Whilst is more performant than a straight loop?

Reply:
I am very convinced it is. It does some performance optimisations under the hood and definitely doesn't block the event loop.